### PR TITLE
Registry migration tokens map refactor

### PR DESCRIPTION
**Pull request by ug02fast**
Removing tokens_map and dependencies of it.

branched off swap refactor. swap refactor should go first.

tested swapping locally.

### CODE REVIEW

**Review by nguyenaiden — APPROVED** [review]
+1 on this. Also, reiterating to move towards default state to reduce null/undefined checks

**Comment by hammeiam** [comment]
fyi you can specify your root branch. So if your branches are `master -> A -> B`, you can set your root to A instead of master and github will only display the diff between A and B. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

File: src/utils/updateTokenPrices.ts [hunk 1]
```diff
@@ -25,79 +25,13 @@ interface CoinGeckoReponse {
     usd: number
   }
 }
-const otherTokens = {
```
Comment (hammeiam) [inline src/utils/updateTokenPrices.ts:28]: I think it would be better to move the changes to this file + App.ts to another PR, so we can revert just that if something ends up going wrong once the new price fetching logic is engaged.

File: src/utils/__tests__/index.ts [hunk 2]
```diff
@@ -207,8 +193,10 @@ describe("getTokenIconPath", () => {
     })
   })
   it("correctly retrieves icon path for saddle tokens", () => {
-    Object.keys(TOKENS_MAP).forEach((tokenSymbol) => {
+    Object.keys({ "saddle-lp": { address: "0x" } }).forEach((tokenSymbol) => {
+      console.log({ tokenSymbol })
```
Comment (hammeiam) [inline src/utils/__tests__/index.ts:197]: rm console logs

File: src/constants/index.ts [hunk 3]
```diff
@@ -2060,33 +2061,13 @@ export type TokensMap = {
   [symbol: string]: Token
 }
 
-export const TOKENS_MAP = Object.keys(POOLS_MAP).reduce((acc, poolName) => {
```
Comment (hammeiam) [inline src/constants/index.ts:2063]: amaaaaaazing

File: src/hooks/useContract.ts [hunk 4]
```diff
@@ -342,13 +342,14 @@ interface AllContractsObject {
 }
 export function useAllContracts(): AllContractsObject | null {
```
Comment (hammeiam) [inline src/hooks/useContract.ts:343]: Can you remove this function too? It's only used in `useApproveAndSwap`, and only to get erc20 contracts for approval.

File: src/pages/App.tsx [hunk 5]
```diff
@@ -198,8 +200,8 @@ function PricesAndVoteData({
     void fetchGasPrices(dispatch)
   }, [dispatch])
   const fetchAndUpdateTokensPrice = useCallback(() => {
-    fetchTokenPricesUSD(dispatch, sdlWethSushiPool, chainId)
-  }, [dispatch, chainId, sdlWethSushiPool])
+    void getTokenPrice(dispatch, tokens, chainId, sdlWethSushiPool)
```
Comment (hammeiam) [inline src/pages/App.tsx:203]: mentioned elsewhere, let's save turning this on for another PR

File: src/hooks/usePendingSwapData.ts [hunk 6]
```diff
@@ -517,11 +521,8 @@ async function fetchPendingSwapInfo(
     } else {
       swapType = SWAP_TYPES.INVALID
     }
-    const synthTokenFrom = getTokenByAddress(
-      pendingSwapInfo.synth,
-      chainId,
-    ) as Token
-    const tokenTo = getTokenByAddress(pendingSwapInfo.tokenTo, chainId) as Token
+    const synthTokenFrom = tokens?.[pendingSwapInfo.synth]
+    const tokenTo = tokens?.[pendingSwapInfo.tokenTo]
```
Comment (hammeiam) [inline src/hooks/usePendingSwapData.ts:525]: it's okay to return null here if either of these tokens are undefined. Just make sure to `console.error` it. This means that you no longer have to add `undefined` to the `PendingSwap` type

File: src/hooks/useTokenMaps.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { BasicTokensMap, PoolName, TokenToPoolsMap } from "../constants"
+
+import { BasicPoolsContext } from "../providers/BasicPoolsProvider"
+import { TokensContext } from "../providers/TokensProvider"
+import { useContext } from "react"
+
+export const useTokenMaps = (): {
```
Comment (hammeiam) [inline src/hooks/useTokenMaps.ts:7]: You can simplify this whole file once https://github.com/saddle-finance/saddle-frontend/pull/1193 lands

File: src/hooks/useTokenMaps.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  tokensMap: BasicTokensMap
+  tokenToPoolsMap: TokenToPoolsMap
+} => {
+  const basicPools = useContext(BasicPoolsContext)
+  const tokens = useContext(TokensContext)
+
+  const tokensMap = Object.keys(basicPools ?? {}).reduce((acc, poolName) => {
+    const pool = basicPools?.[poolName as PoolName]
```
Comment (hammeiam) [inline src/hooks/useTokenMaps.ts:15]: `if !pool return acc`, get rid of `pool?.`

File: src/hooks/useTokenMaps.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  const tokenToPoolsMap = Object.keys(basicPools ?? {}).reduce(
+    (acc, poolName) => {
+      const pool = basicPools?.[poolName]
+      const poolTokens = pool?.tokens.map((token) => tokens?.[token]) ?? []
+      const newAcc = { ...acc }
+      poolTokens.forEach((token) => {
+        newAcc[token?.symbol ?? ""] = (
+          newAcc[token?.symbol ?? ""] || []
+        ).concat(poolName as PoolName)
```
Comment (hammeiam) [inline src/hooks/useTokenMaps.ts:35]: kill PoolName

File: src/hooks/useTokenMaps.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    newAcc[lpTokenSymbol] = lpToken
+    return newAcc
+  }, {} as BasicTokensMap)
+
+  const tokenToPoolsMap = Object.keys(basicPools ?? {}).reduce(
+    (acc, poolName) => {
+      const pool = basicPools?.[poolName]
+      const poolTokens = pool?.tokens.map((token) => tokens?.[token]) ?? []
+      const newAcc = { ...acc }
+      poolTokens.forEach((token) => {
```
Comment (hammeiam) [inline src/hooks/useTokenMaps.ts:32]: `if !token return`, skip all the `token?.symbol ?? ""` later

File: src/pages/Swap.tsx [hunk 11]
```diff
@@ -128,11 +127,11 @@ function Swap(): ReactElement {
         to: [],
       }
 
-    const allTokens = Object.values(TOKENS_MAP)
-      .filter(({ isLPToken, addresses }) => !isLPToken && addresses[chainId])
+    const allTokens = Object.values(tokens || {})
+      .filter(({ isLPToken, address }) => !isLPToken && address)
```
Comment (hammeiam) [inline src/pages/Swap.tsx:131]: you can remove `address` here, the old logic was just making sure the token existed on the current chain, but `tokens` does that automatically.

File: src/pages/Swap.tsx [hunk 12]
```diff
@@ -163,21 +162,21 @@ function Swap(): ReactElement {
         ? (
             formState.currentSwapPairs
               .map(({ to, type: swapType }) => {
-                if (!TOKENS_MAP?.[to.symbol]) {
+                if (!tokensMap?.[to.symbol]) {
```
Comment (hammeiam) [inline src/pages/Swap.tsx:165]:
```
const token = tokensMap?.[to.symbol]
if (!token) log and return null
```
now token is guaranteed to be defined and you can skip all of the `token?.name ?? ""` again

File: src/pages/Swap.tsx [hunk 13]
```diff
@@ -248,11 +250,14 @@ function Swap(): ReactElement {
...
       ) {
-        const originPool = POOLS_MAP[formStateArg.from.poolName]
-        const destinationPool = POOLS_MAP[formStateArg.to.poolName]
+        const originPool = basicPools?.[formStateArg.from.poolName]
+        const destinationPool = basicPools?.[formStateArg.to.poolName]
         const [amountOutSynth, amountOutToken] =
           await bridgeContract.calcTokenToToken(
-            [originPool.addresses[chainId], destinationPool.addresses[chainId]],
+            [
+              originPool?.metaSwapDepositAddress ?? "",
```
Comment (hammeiam) [inline src/pages/Swap.tsx:258]: this will fail if we don't know the address. Write to the `error` var and set `amountToReceive, amountMediumSynth` to `Zero` instead

File: src/pages/Swap.tsx [hunk 14]
```diff
@@ -263,10 +268,10 @@ function Swap(): ReactElement {
         formStateArg.swapType === SWAP_TYPES.SYNTH_TO_TOKEN &&
         bridgeContract != null
       ) {
-        const destinationPool = POOLS_MAP[formStateArg.to.poolName]
+        const destinationPool = basicPools?.[formStateArg.to.poolName]
         const [amountOutSynth, amountOutToken] =
           await bridgeContract.calcSynthToToken(
-            destinationPool.addresses[chainId],
+            destinationPool?.metaSwapDepositAddress ?? "",
```
Comment (hammeiam) [inline src/pages/Swap.tsx:274]: same as above

File: src/pages/Swap.tsx [hunk 15]
```diff
@@ -277,9 +282,9 @@ function Swap(): ReactElement {
         formStateArg.swapType === SWAP_TYPES.TOKEN_TO_SYNTH &&
         bridgeContract != null
       ) {
-        const originPool = POOLS_MAP[formStateArg.from.poolName]
+        const originPool = basicPools?.[formStateArg.from.poolName]
         amountToReceive = await bridgeContract.calcTokenToSynth(
-          originPool.addresses[chainId],
+          originPool?.metaSwapDepositAddress ?? "",
```
Comment (hammeiam) [inline src/pages/Swap.tsx:287]: same as above

File: src/pages/Swap.tsx [hunk 16]
```diff
@@ -232,10 +234,10 @@ function Swap(): ReactElement {
       }
       const amountToGive = parseUnits(
         cleanedFormFromValue,
-        TOKENS_MAP[formStateArg.from.symbol].decimals,
+        tokensMap?.[formStateArg.from.symbol]?.decimals ?? 0,
       )
-      const tokenFrom = TOKENS_MAP[formStateArg.from.symbol]
-      const tokenTo = TOKENS_MAP[formStateArg.to.symbol]
+      const tokenFrom = tokensMap[formStateArg.from.symbol]
```
Comment (hammeiam) [inline src/pages/Swap.tsx:239]: hoist both of these to the top of the function. I think we should either assume that if these tokens are in `formStateArg`, then they must also be in `tokensMap`, then we can force them to be `BasicToken` and not undefined. Or we check if they're undefined around line 213 and not even try to use them

File: src/pages/Swap.tsx [hunk 17]
```diff
@@ -325,9 +330,9 @@ function Swap(): ReactElement {
           priceImpact,
           exchangeRate: calculateExchangeRate(
             amountToGive,
-            tokenFrom.decimals,
+            tokenFrom?.decimals ?? 0,
             amountToReceive,
-            tokenTo.decimals,
+            tokenTo?.decimals ?? 0,
```
Comment (hammeiam) [inline src/pages/Swap.tsx:335]: Prefer checking for token existence instead of making up numbers

File: src/pages/Swap.tsx [hunk 18]
```diff
@@ -578,7 +583,7 @@ function Swap(): ReactElement {
             ? "0"
             : formatUnits(
                 formState.to.value,
-                TOKENS_MAP[formState.to.symbol].decimals,
+                tokensMap[formState.to.symbol]?.decimals ?? 18,
```
Comment (hammeiam) [inline src/pages/Swap.tsx:586]: same as above, prefer not to make up numbers.

File: src/hooks/usePendingSwapData.ts [hunk 19]
```diff
@@ -20,9 +23,9 @@ export interface PendingSwap {
   swapType: SWAP_TYPES
   settleableAtTimestamp: number
   secondsRemaining: number
-  synthTokenFrom: Token
+  synthTokenFrom: BasicToken | undefined
```
Comment (hammeiam) [inline src/hooks/usePendingSwapData.ts:26]: see comment below

File: src/components/PendingSwapExchange.tsx [hunk 20]
```diff
@@ -38,20 +38,20 @@ const PendingSwapExchange = ({
   const { t } = useTranslation()
   const { synthTokenFrom, tokenTo, synthBalance, swapType } = pendingSwap
   const formattedSynthBalance = commify(
-    formatBNToString(synthBalance, synthTokenFrom.decimals, 6),
+    formatBNToString(synthBalance, synthTokenFrom?.decimals ?? 0, 6),
```
Comment (hammeiam) [inline src/components/PendingSwapExchange.tsx:41]: these changes are all moot if you don't add `| undefined` to `PendingSwap` type, see comments in `usePendingSwapData`

File: src/components/PendingSwapModal.tsx [hunk 21]
```diff
@@ -194,7 +194,7 @@ const PendingSwapModal = ({
       ? synthBalance
       : settlementState.amount
   const formattedSynthBalance = commify(
-    formatBNToString(synthBalance, synthTokenFrom.decimals, 6),
+    formatBNToString(synthBalance, synthTokenFrom?.decimals ?? 0, 6),
```
Comment (hammeiam) [inline src/components/PendingSwapModal.tsx:197]: same as `PendingSwapExchange` I believe

File: src/components/PendingSwapTimeline.tsx [hunk 22]
```diff
@@ -41,7 +41,7 @@ const PendingSwapTimeline = ({
   } = pendingSwap
   const minutesRemaining = Math.max(Math.ceil(secondsRemaining / 60), 0)
   const formattedBalance = commify(
-    formatBNToString(synthBalance, synthTokenFrom.decimals, 6),
+    formatBNToString(synthBalance, synthTokenFrom?.decimals ?? 0, 6),
```
Comment (hammeiam) [inline src/components/PendingSwapTimeline.tsx:44]: same as `PendingSwapExchange`

File: src/components/SwapPage.tsx [hunk 23]
```diff
@@ -259,7 +259,7 @@ const SwapPage = (props: Props): ReactElement => {
           const formattedSynthBalance = commify(
             formatUnits(
               pendingSwap.synthBalance,
-              pendingSwap.synthTokenFrom.decimals,
+              pendingSwap.synthTokenFrom?.decimals ?? 0,
```
Comment (hammeiam) [inline src/components/SwapPage.tsx:262]: same as `PendingSwapExchange`

**Review by hammeiam — CHANGES_REQUESTED** [review]
tldr assert that a token exists once by early returning, save yourself from having to write `token?.symbol ?? ""` in 8 different files

File: src/hooks/usePendingSwapData.ts [hunk 24]
```diff
@@ -335,7 +339,7 @@ async function fetchAndPopulatePendingSwaps(
 
   // Step 3: fetch aync data via bridge contract
   let fetchedPendingSwaps = await Promise.all(
```
Comment (nguyenaiden) [inline src/hooks/usePendingSwapData.ts:341]: While you're at it, could you take a look and change these to
```
const events = (
    await Promise.all(
      eventFilters.map((filter) =>
        bridgeContract.queryFilter(filter, startBlock, "latest"),
      ),
    )
  ).flat()

  // Step 3: fetch aync data via bridge contract
  const fetchedPendingSwaps = (
    await Promise.all(
      events.map((event) =>
        fetchPendingSwapInfo(bridgeContract, event, chainId),
      ),
    )
  ).filter(Boolean)
  if (fetchedPendingSwaps.length === 0) return
  ```
Reassignments are usually not a good idea.

Comment (ug02fast) [reply src/hooks/usePendingSwapData.ts:341]: This is a good call but can you make an issue on this and make the change in a separate PR for the sake of keeping the scope of this to removal of POOL/TOKENS_MAP?

File: src/utils/updateTokenPrices.ts (hunk 1 above)
Comment (ug02fast) [reply src/utils/updateTokenPrices.ts:28]: Token price fetching is working with the changes in this PR. What do you think would be a reason for something to go wrong?

File: src/pages/Swap.tsx (hunk 18 above)
Comment (ug02fast) [reply src/pages/Swap.tsx:609]: wasnt quiet able to figure this one out.

**Review by nguyenaiden — APPROVED** [review]
Okay this looks good to continue with the migration. I've manually tested deposit, swap, and withdraw locally.

Good to merge after addressing a couple of comments

Will need to create tickets to go back and address setting default states to allow early exits and remove optional. High priority

File: src/hooks/useCalculateSwapPairs.ts [hunk 25]
```diff
@@ -99,11 +99,11 @@ export function useCalculateSwapPairs(): (token?: Token) => SwapData[] {
   }, [chainId])
 
   return useCallback(
-    function calculateSwapPairs(token?: Token): SwapData[] {
+    function calculateSwapPairs(token?: BasicToken): SwapData[] {
       if (token == null || chainId == null || tokens === null) return []
       const cacheHit = pairCache[token.symbol]
       if (cacheHit) return cacheHit
-      const originTokenAddress = (token.addresses[chainId] || "").toLowerCase()
+      const originTokenAddress = (token?.address || "").toLowerCase()
```
[the remark this answers is not in the document]
Comment (ug02fast) [reply src/hooks/useCalculateSwapPairs.ts:106]: done in following PR https://github.com/saddle-finance/saddle-frontend/pull/1182/files#diff-ab89847bbe63169d79f42de9dbaaac7dd80b8651f8bd33a3a4e7f523d34c191bR106

File: src/hooks/useCalculateSwapPairs.ts [hunk 26]
```diff
@@ -99,11 +99,11 @@ export function useCalculateSwapPairs(): (token?: Token) => SwapData[] {
   }, [chainId])
 
   return useCallback(
-    function calculateSwapPairs(token?: Token): SwapData[] {
+    function calculateSwapPairs(token?: BasicToken): SwapData[] {
       if (token == null || chainId == null || tokens === null) return []
```
[the remark this answers is not in the document]
Comment (ug02fast) [reply src/hooks/useCalculateSwapPairs.ts:103]: For the sake of keeping the scope of PR from pulling out TOKENS_MAP going to leave as is. It would be better to discuss conventions as a team and make one PR to replace everywhere.

File: src/hooks/usePendingSwapData.ts [hunk 27]
```diff
@@ -492,15 +496,15 @@ enum BridgePendingSwapTypes {
...
+  tokens: BasicTokens,
 ) {
   let result = null
   try {
     // this will throw if the itemId has already fully resolved
     // afik we don't have a better way of fetching current pendingSwaps
     // DEVELOPER: for this reason we don't use multicall
     const pendingSwap = event.args as unknown as BridgeEventGenericSwap
-    if (pendingSwap == null) return null
+    if (pendingSwap == null || !tokens) return null
```
[the remark this answers is not in the document]
Comment (ug02fast) [reply src/hooks/usePendingSwapData.ts:507]: For the sake of keeping the scope of PR from pulling out TOKENS_MAP going to leave as is. It would be better to discuss conventions as a team and make one PR to replace everywhere.

File: src/hooks/useApproveAndSwap.ts [hunk 28]
```diff
@@ -65,18 +67,26 @@ export function useApproveAndSwap(): (
...
       if (state.swapType === SWAP_TYPES.DIRECT && !state.swapContract)
         throw new Error("Swap contract is not loaded")
       if (state.swapType !== SWAP_TYPES.DIRECT && !state.bridgeContract)
         throw new Error("Bridge contract is not loaded")
       if (chainId === undefined) throw new Error("Unknown chain")
       // For each token being deposited, check the allowance and approve it if necessary
-      const tokenContract = tokenContracts?.[state.from.symbol] as Erc20
+      const token = tokensMap[state.from.symbol]
+      if (!token) throw new Error("Token not found")
+      const tokenContract = getContract(
```
Comment (hammeiam) [inline src/hooks/useApproveAndSwap.ts:84]: Awesome! This is so much better. Nit, no need to lowercase address here. The only reason we do that elsewhere is to make addresses uniform for equality checks
